### DISCUSSION TITLE
v8.4.0 Deprecation Notice

### DISCUSSION THREAD

**Discussion by BottlecapDave** [BottlecapDave/HomeAssistant-OctopusEnergy#444] [Announcements]
Hello and thank you for using my integration.

With the release of v8.4.0, I've introduced [events](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/events.md) being raised by the integration when various rate information is updated. The main reason for this is to power some new `event` sensors which is the new home for accessing raw rate data. I've updated the [community](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/community.md) contributions to reflect this change.

The main purpose for this is to go towards fixing https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/issues/357. Therefore, the following attributes are now deprecated and will be removed in the next major version which will be released after 07/11/2023. 

`sensor.octopus_energy_electricity_{{METER_SERIAL_NUMBER}}_{{MPAN_NUMBER}}_current_rate`
* `all_rates` - This data has been split across the new [current](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#current-day-rates), [previous ](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#previous-day-rates) and [next](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#next-day-rates) day event sensors
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

`sensor.octopus_energy_electricity_{{METER_SERIAL_NUMBER}}_{{MPAN_NUMBER}}_previous_rate`
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

`sensor.octopus_energy_electricity_{{METER_SERIAL_NUMBER}}_{{MPAN_NUMBER}}_next_rate`
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

`sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_current_rate`
* `all_rates` - This data has been split across the new [current](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/gas.md#current-day-rates), [previous ](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/gas.md#previous-day-rates) and [next](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/gas.md#next-day-rates) day event sensors
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

`sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_previous_rate`
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

`sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_next_rate`
* `applicable_rates` - This will no longer be available. All data can be derrived from the other attributes and was used more for debugging purposes.

**Comment by willenic** [comment]
Hi Dave,

Just getting my head round this in advance.  So under the next release we'll need to check the rates attribute and compare current time/or desired time to perform an action with the time in rates attribute list for current/next day to perform automations/calculations against?

Cheers,

Will

**Reply by BottlecapDave** [reply]
If you're not using the mentioned attributes within the mentioned sensors, then you will not need to do anything as everything else will continue to work.

If you are using the `all_rates` attribute (e.g. for graphs), then you'll need to migrate to use the new event entities. As mentioned, I've already updated the community graphs to use these new sensors.

If you are using the deprecated `applicable_rates` attribute, I'd like to know what you're using it for to better help with the migration.

**Reply by willenic** [reply]
Got, it thought you were deprecating the sensor, misread on my phone on the train. Cheers.

**Reply by willenic** [reply]
@BottlecapDave I'm currently comparing the state of current and next rate entities to various input variables to turn on and off my Hot Water, Dishwasher and car charger. 

I don't use the targets because all of these (and no doubt future devices on the to do list) have different delays between calling the service and the desired device being turned on. Additionally, the car and hot water are more variable lengths of time so I'd need many target sensors with different offsets mixed between continuous and intermittent.

It seemed simpler to compare the state of ```sensor.octopus_energy_electricity_{{METER_SERIAL_NUMBER}}_{{MPAN_NUMBER}}_next_rate``` with my input variable ```desired rate_to_call_service``` which will continue to work but is slightly inelegant when the next rate lasts for more than 30 minutes.

Using the event sensors would get me out of handling exceptions when next_rate state is unknown though, I just haven't found out how to select the ```value_inc_vat``` from the nth record in the list which is what I'm struggling with.

**Reply by BottlecapDave** [reply]
> I don't use the targets because all of these (and no doubt future devices on the to do list) have different delays between calling the service and the desired device being turned on.

Target rate sensors support this with the [offset](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md#offset) parameter. This allows you to offset when the sensor comes on (e.g. 30 minutes before/after the best slot), therefore adding a delay

>Additionally, the car and hot water are more variable lengths of time so I'd need many target sensors with different offsets mixed between continuous and intermittent.

A lot of people use the [update config](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/services.md#service-octopus_energyupdate_target_config) service to adjust target rates throughout the day based on outside parameters. So you can use this to change the target hours dynamically.

> Using the event sensors would get me out of handling exceptions when next_rate state is unknown though, I just haven't found out how to select the value_inc_vat from the nth record in the list which is what I'm struggling with.

The rate attribute is just a normal array, so you'd access the record like a normal array

```
{{ state_attr('event.xxx', 'rates')[0]["value_inc_vat"] }}
```

if you wanted to access hour 3 and minute 30, then the index would be `(hour (e.g. 3) x 2) + (minute (e.g. 32) % 30)` which would equal 7.

**Reply by willenic** [reply]
Thanks, as ever. I'll have a look at the update config service which will probably get me what I need.

**Comment by springfall2008** [comment]
Hi Dave,

I'm using 'all_rates' in these sensors, what will be the new way to get the same rate data in HA?

Thanks

**Reply by springfall2008** [reply]
Oh sorry I see it now 'This data has been split across the new [current](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/sensors/electricity.md#current-day-rates), [previous ](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/sensors/electricity.md#previous-day-rates)and [next](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/sensors/electricity.md#next-day-rates) day event sensors'

**Comment by JRascagneres** [comment]
FYI - the links to the new sensors are broken

**Reply by BottlecapDave** [reply]
I was rearranging the docs - thanks for the heads up. These should now be fixed.

**Comment by springfall2008** [comment]
>all_rates - This data has been split across the new [current](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#current-day-rates), [previous ](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#previous-day-rates)and [next](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#next-day-rates) day event sensors

I just installed v9.0.0 and was expecting all_rates to still work, but it seems to be gone and I can't find the Agile rates anymore. I need the next 24-48 hours of rate data for Predbat.

**Reply by JRascagneres** [reply]
I think that sensor was removed due to it being too large for HA to be happy with. I think you need to join the day and next day

**Reply by springfall2008** [reply]
Neither day or next day seem to have the half hourly Agile rates

**Reply by JRascagneres** [reply]
I would check but I'm getting solar installed as we speak so my HA is down... I'm fairly sure mine had it tho...

**Reply by BottlecapDave** [reply]
This post made it clear that the mentioned attributes would be removed in the next major release to fix the forementioned issue. If you have only just enabled the event entities, it might take up to 30 minutes to populate while they wait for the event to fire. If you want to speed this process up, a reload of the integration should speed it up.

**Reply by springfall2008** [reply]
@BottlecapDave ah okay you are saying this has moved to different entities for events, let me have a look?

**Reply by JRascagneres** [reply]
Works for me
<img width="814" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/79d12a32-6e86-44cc-adda-674b588fed49">

**Reply by deltamelter** [reply]
where are you able to see the "state attributes" listed like this?  the sensor had the drop down but as an event I can't find the attributes.  They are there on the graph and usable in code, but like to see the "raw" data like this when I am creating new graphs or javascript cards.

**Reply by deltamelter** [reply]
nvm, found the states tab on developer tools.

**Reply by UK-JustinSmall** [reply]
Any idea how to include this in apex charts. I have the following snipit.

series:
  - entity: sensor.octopus_energy_electricity_AAAAAAA_BBBBBBBBB_current_rate
    type: column
    name: price
    color: gray
    opacity: 1
    stroke_width: 0
    unit: p/Kw

Tried changing it to 

 - entity: sensor.octopus_energy_electricity_AAAAAAA_BBBBBBBBB_current_day_rates

**Reply by noifen** [reply]
@UK-JustinSmall you need to change it to `event.,` not `sensor.`

Got a post on reddit for more sample config https://www.reddit.com/r/OctopusEnergy/comments/180etza/ha_integration_v90/

**Reply by BottlecapDave** [reply]
The [community examples](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/community.md) have been updated based on the new structure.

**Comment by fboundy** [comment]
They have been in events for some time. The event entities are disabled by default so you need to enable them manually

**Reply by springfall2008** [reply]
Just trying to enable them now, this is a nightmare support wise getting people to manually enable stuff. I wonder if the rates could be enabled by a configuration option for the plugin instead what do you think @BottlecapDave ?

**Reply by JRascagneres** [reply]
tbh I think they should be auto enabled once the other sensor is removed

**Reply by BottlecapDave** [reply]
They were disabled by default to not overwhelm new users with lots of enabled entities. However, I'll enable these by default in the next release.

**Reply by springfall2008** [reply]
That would be awesome, thanks for your continued support

**Comment by noifen** [comment]
Not sure if this is an ApexCharts/HA events thing, but the data generator in Apex Charts needs to use `start` rather than `valid_from`. Not sure if that is controlled by the integration or not? Just threw me off for a while making the move.

Full v9 working data generator code:
```
return entity.attributes.rates.map((entry) => {
  return [new Date(entry.start), entry.value_inc_vat*100];
});
```

**Reply by BottlecapDave** [reply]
Thanks. I've updated the community examples.

**Comment by WhyAydan** [comment]
Could be me being thick (most likely) but how do we get the agile rates card to work now? tried using events. but it just shows nothing.

Also noticed the current_rate triggers every minute so my push notifications to tell me the rate real time (30 mins) is firing every minute.

**Reply by deltamelter** [reply]
need to change the period_from to start and the array is called rates now instead of all_rates and lastly the rates are now in decimal GBP so you might not see anything on a graph if there is anything else that has dominant scale

**Reply by WhyAydan** [reply]
@deltamelter do you have the agile rates card config to hand? This is what I have.


title: Agile Octopus Rates
entity: event.octopus_energy_electricity_next_day_rates
type: custom:octopus-energy-rates-card
cols: 2
showday: true
showpast: false
mediumlimit: 25
highlimit: 37

**Reply by deltamelter** [reply]
config alone won't help, there are a few lines in the javascript to change too
also the entity name should have the meter details in it...

```
entity: event.octopus_energy_electricity_XXXXXXXXXXXX_0000000000000_current_day_rates
type: custom:my-octopus-energy-rates-card
cols: 3
showday: true
showpast: true
#medlow: 10
mediumlimit: 20
highlimit: 30
#vhighlimit: 50
```
I see somebody had a go at fixing the card on github, so it will be updated soon but if you are brave enough to wield a text editor change some lines in the javascript
https://github.com/stevetrease/octopus-energy-rates-card/blob/main/octopus-energy-rates-card.js

**Reply by UK-JustinSmall** [reply]
I've gone in and amended the .js file in config/www/community folder with the contents of the javascript file.  https://github.com/stevetrease/octopus-energy-rates-card/blob/main/octopus-energy-rates-card.js

I still get the error
![Snag_1cf8214](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/129121587/452711f4-685f-4d35-838b-fe9ddc64c3d9)

**Comment by deltamelter** [comment]
I know that the rates have been split across previous, current and next day, but is there no overlap?  Does that mean that I need to use 2 sensors instead of 1?  I didn't enable next day until now when 1600 rolled around and the charts and cards didn't reflect the next bunch of prices.  Is there a way to get the next day to update now or do I have to wait until 16:30...   I guess I'll find outin 9 mins :grin:

**Reply by BottlecapDave** [reply]
The events are fired regardless of the entities being enabled. They are fired whenever the rates are refreshed, which is scheduled for every 30 minutes as long as the APIs are available. However data won't be available in the event entities when you first enable them as they need to wait for the above events to fire. You can force this by reloading the integration.

**Reply by BottlecapDave** [reply]
There is no overlap between current and next day event entities. If you are using Apex charts, you can see the change in the community document linked above. If you are using the agile cards, you'll need to wait for an update.

**Reply by deltamelter** [reply]
I reloaded the integration after enabling "next day", but didn't populate until 16:30.
Yes I have a bunch of apex charts which all used the all_rates array.  I'll update them with both entities.   
Does the previous day contain the rates from 23:30 to  23:00 from the previous day then?  I guess I will enable that too as I might need to refer back

**Reply by BottlecapDave** [reply]
Previous day should be 00:00 yesterday to 00:00 today.

**Reply by deltamelter** [reply]
previous day rates has 21:00 to 00:00 only

**Reply by BottlecapDave** [reply]
hmm that's a bug. I'll see if I can fix that.

**Reply by BottlecapDave** [reply]
I've released 9.0.1 which should include the full previous day for agile tariffs.
